src/Core/Config/Util/ConfigFileManager.php

```diff
@@ -144,82 +149,153 @@ private function loadStaticConfig(string $name): array
+					}
 				}
+			} finally {
+				// unlock and close the stream for every circumstances
 				flock($configStream, LOCK_UN);
+				fclose($configStream);
 			}
 
-			fclose($configStream);
-
+			/**
+			 * Evaluate the fetched content
+			 *
+			 * @note
+			 * Because `eval()` directly evaluates PHP content, we need to "close" the expected PHP content again with
+			 * the prefixed "?>". Now we're in plain HTML again and can evaluate any PHP file :-)
+			 */
 			$dataArray = eval('?>' . $content);
 
-			if (!is_array($dataArray)) {
-				throw new ConfigFileException(sprintf('Error loading config file %s', $filename));
+			if (is_array($dataArray)) {
+				$configCache->load($dataArray, Cache::SOURCE_DATA);
 			}
-
-			$config->load($dataArray, Cache::SOURCE_DATA);
 		}
 	}
 
 	/**
-	 * Saves overridden config entries back into the data.config.phpR
+	 * Saves overridden config entries back into the data.config.php
 	 *
-	 * @param Cache $config The config cache
+	 * @param Cache $configCache The config cache
 	 *
 	 * @throws ConfigFileException In case the config file isn't writeable or the data is invalid
 	 */
-	public function saveData(Cache $config)
+	public function saveData(Cache $configCache)
 	{
-		$data = $config->getDataBySource(Cache::SOURCE_DATA);
-
-		$encodedData = ConfigFileTransformer::encode($data);
+		$filename = $this->configDir . '/' . self::CONFIG_DATA_FILE;
 
-		if (!$encodedData) {
-			throw new ConfigFileException('config source cannot get encoded');
+		if (file_exists($filename)) {
+			$fileExists = true;
+		} else {
+			$fileExists = false;
 		}
 
-		$configStream = fopen($this->configDir . '/' . self::CONFIG_DATA_FILE, 'w');
+		/**
+		 * Creates a read-write stream
+		 *
+		 * @see https://www.php.net/manual/en/function.fopen.php
+		 * @note Open the file for reading and writing. If the file does not exist, it is created.
+		 * If it exists, it is neither truncated (as opposed to 'w'), nor the call to this function fails
+		 * (as is the case with 'x'). The file pointer is positioned on the beginning of the file.
+		 *
+		 */
+		$configStream = fopen($filename, 'c+');
+
+		try {
+			// We do want an exclusive lock, so we wait until every LOCK_SH (config reading) is unlocked
+			if (flock($configStream, LOCK_EX)) {
+
+				/**
+				 * If the file exists, read the whole file again
+				 * Since we're currently exclusive locked, no other process can now change the config again
+				 */
+				if ($fileExists) {
+					// When reading the config file too fast, we get a wrong filesize, "clearstatcache" prevents that
+					clearstatcache(true, $filename);
+					$content = fread($configStream, filesize($filename));
+					// Event truncating the whole content wouldn't automatically rewind the stream,
+					// so we need to do it manually
+					rewind($configStream);
+					if (!$content) {
+						throw new ConfigFileException(sprintf('Cannot read file %s', $filename));
+					}
```

Comment:
Is it meaningful to rewind the pointer before checking for content?
```suggestion
					if (!$content) {
						throw new ConfigFileException(sprintf('Cannot read file %s', $filename));
					}

					// Event truncating the whole content wouldn't automatically rewind the stream,
					// so we need to do it manually
					rewind($configStream);
```